File: 002/Python/problem002.py
def fib_sum(n):
    return sum(filter(even, fibs_upto(n)))

def fibs_upto(n):
    if n == 0:
        return []
    else:
        return fibs_upto_helper(n, 2, 1, [])
        
def fibs_upto_helper(n, first, second, fibs):
    if first > n:
        return [second] + fibs
    else:
        return fibs_upto_helper(n, first + second, first, [second] + fibs)

def even(n):
    return n % 2 == 0

File: 002/Ruby/problem002.rb
class Integer
  def fib_sum
    fibs_upto.select { |n| n.even? }.inject(0) { |sum, n| sum + n }
  end
  
  def fibs_upto
    if zero?
      []
    else
      fibs_upto_helper(2, 1, [])
    end
  end
  
  private
  
  def fibs_upto_helper(first, second, fibs)
    if first > self
      [second] + fibs
    else
      fibs_upto_helper(first + second, first, [second] + fibs)
    end
  end
end
